Renew DHCPv4 address if network connectivity is restored

jukkar opened the pull request:
Fixes #7553

jukkar committed:
drivers: eth: mcux: Inform IP stack when carrier is lost

If carrier is ON or OFF, then tell this information to upper IP stack
so that it can act accordingly.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: dhcpv4: Fix IPv4 and UDP checksum calculation

Make sure we calculate the IPv4 and UDP checksum only when
needed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: dhcpv4: Detect network interface on/off events

Catch interface up/down events so that we can renew the
address if interface goes down and is then restored.

Fixes #7553

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka commented on include/net/net_if.h:
```diff
@@ -1600,6 +1600,14 @@ static inline bool net_if_is_up(struct net_if *iface)
+ *
+ * @param iface Pointer to network interface
+ */
+void net_if_carrier_down(struct net_if *iface);
```
huh no, carrier is really something ethernet centric. Actually see net/ethernet_mgmt.h there are ethernet_mgmt_raise_carrier_<on/off>_event() functions

that said, for a more generic stuff we could have a link_up and link_down as status flag.

However, that requires a bit of thought to get this right. We had that discussion already back in the days with @Vudentz when he introduced up/down status flag and getting all these states proper would need some changes. up/down would be about IP core being ready/off on the iface. link up/down would be about the hw being ready (for instance, in case of ethernet there would no reason to put the link up if the carrier is off).  etc...

jukkar replied:
Hmm, net_if_carrier_down() is "generic" and already used by Bluetooth (see subsys/net/ip/l2/bluetooth.c)

tbursztyka replied:
that makes sense because before being able to do ipsp, you need an actual bt connection below.
Like in ethernet you need an actual cable being plugged in (carrier line).

So carrier is still specific to certain bearers then. There is no such things in 15.4 for instance or slip tap etc...

I can't recall exactly what was the discussion, it was on gerrit back then (if I remember well) so that's a long time ago. Idea was the we need to settle different status flags for exact proper state of device/iface so it's not messed up through device drivers, l2 and ip core stack.

Well, we probably don't have time for 1.12 to think that through so let's make it easier, it will be already an improvement: add net_if_carrier_up(), the Ethernet device driver would call that one. In net_if.c then call net_if_up() there. This will enable future features like not automatically "connect" (putting the iface up with an ip etc...) the device.

tbursztyka commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -286,12 +286,14 @@ static void eth_mcux_phy_event(struct eth_context *context)
 					  kENET_MiiReadValidFrame);
 			context->link_up = link_up;
 			context->phy_state = eth_mcux_phy_state_read_duplex;
+			net_if_up(context->iface);
```
so following my other comment, that should never be raised by the driver. (link up should, as well as carrier mgmt event btw)

jukkar replied:
So your idea is that we raise yet another event here instead of calling the function directly?

tbursztyka commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -1069,11 +1072,39 @@ static enum net_verdict net_dhcpv4_input(struct net_conn *conn,
 	u32_t entropy;
 
+	net_mgmt_init_event_callback(&mgmt4_cb, iface_handler,
+				     NET_EVENT_IF_DOWN | NET_EVENT_IF_UP);
```
change your commit title, you are listening to up/down events, not carrier.

jukkar committed:
net: ethernet: Provide stubs for ethernet carrier functions

If CONFIG_NET_L2_ETHERNET_MGMT is not enabled, then provide
stubs for ethernet_mgmt_raise_carrier_on_event() and
ethernet_mgmt_raise_carrier_off_event() functions so that those
functions can be called always.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

therealprof commented on subsys/net/ip/l2/ethernet/ethernet.c:
```diff
@@ -757,18 +759,43 @@ int net_eth_vlan_disable(struct net_if *iface, u16_t tag)
+
+	NET_DBG("Initializing Ethernet L2 %p for iface %p", ctx, iface);
 
+	if (!ctx->is_init) {
```
Missing `#if defined(CONFIG_NET_L2_ETHERNET_MGMT)` gate here.

tbursztyka commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -286,12 +287,14 @@ static void eth_mcux_phy_event(struct eth_context *context)
 					  kENET_MiiReadValidFrame);
 			context->link_up = link_up;
 			context->phy_state = eth_mcux_phy_state_read_duplex;
+			ethernet_mgmt_raise_carrier_on_event(context->iface);
```
Hum, I am not super found of raising net_mgmt events from device drivers directly. (only in very particular case as in offloaded drivers, and still: it does not do it directly).

Instead, create 2 ethernet public functions implemented in ethernet l2 side:  ethernet_carrier_on()  and ethernet_carrier_off() 
There you could handle what you already do in the l2 side (calling net_if_up(), etc... and raise the events).
At least, form drivers/l2/net_if point of view it's reliable that way. (let's avoid using net_mgmt for core features, unless we are really forced to.)

That said, it helped to find out an issue in net_mgmt: it was not possible to raise an event in ISR mode.

tbursztyka commented on subsys/net/ip/l2/ethernet/ethernet.c:
```diff
@@ -757,18 +759,45 @@ int net_eth_vlan_disable(struct net_if *iface, u16_t tag)
+static void carrier_handler(struct net_mgmt_event_callback *cb,
+			    u32_t mgmt_event, struct net_if *iface)
+{
+	if (mgmt_event == NET_EVENT_ETHERNET_CARRIER_ON) {
```
so move this in a ethernet_carrier_on() function  (also call the net event for ethernet carrier on relevantly)

tbursztyka commented on subsys/net/ip/l2/ethernet/ethernet.c:
```diff
@@ -757,18 +759,45 @@ int net_eth_vlan_disable(struct net_if *iface, u16_t tag)
+		NET_DBG("Carrier ON for Interface %p", iface);
+		net_if_up(iface);
+	} else if (mgmt_event == NET_EVENT_ETHERNET_CARRIER_OFF) {
+		NET_DBG("Carrier OFF for Interface %p", iface);
```
and that to a ethernet_carrier_off() function  (also call the net event for ethernet carrier off relevantly)

tbursztyka commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -1069,11 +1072,40 @@ static enum net_verdict net_dhcpv4_input(struct net_conn *conn,
 	return NET_DROP;
 }
 
+static void iface_handler(struct net_mgmt_event_callback *cb,
```
iface_event_handler() would be better name

tbursztyka commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -1069,11 +1072,40 @@ static enum net_verdict net_dhcpv4_input(struct net_conn *conn,
+	net_mgmt_init_event_callback(&mgmt4_cb, iface_handler,
+				     NET_EVENT_IF_DOWN | NET_EVENT_IF_UP);
+	net_mgmt_add_event_callback(&mgmt4_cb);
+
```
Thing is, it won't work if the cable is not plugged in at start up but plugged later. So before k_delayed_work_submit() it should verify that net_if is up.

and maybe it should register the event handler after initialization.

tbursztyka commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -1069,11 +1072,40 @@ static enum net_verdict net_dhcpv4_input(struct net_conn *conn,
+static void iface_handler(struct net_mgmt_event_callback *cb,
+			  u32_t mgmt_event, struct net_if *iface)
+{
+	if (iface->config.dhcpv4.state == NET_DHCPV4_INIT) {
```
read my comment below about k_delayed_work_submit(), I think you don't need that test then

tbursztyka commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -1069,11 +1072,40 @@ static enum net_verdict net_dhcpv4_input(struct net_conn *conn,
+		NET_DBG("Interface %p going down", iface);
+
+		iface->config.dhcpv4.attempts = 0;
+		iface->config.dhcpv4.state = NET_DHCPV4_RENEWING;
```
should it be renewing if only the state was BOUND before?

tbursztyka commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -1069,11 +1072,40 @@ static enum net_verdict net_dhcpv4_input(struct net_conn *conn,
+	} else if (mgmt_event == NET_EVENT_IF_UP) {
+		NET_DBG("Interface %p coming up", iface);
+
+		if (iface->config.dhcpv4.state == NET_DHCPV4_RENEWING) {
```
(following my comment about k_delayed_work_submitg below) this should check for INIT as well then.

tbursztyka reviewed (changes requested):
Various changes requested.

Also, while testing on top of #8112, I still fail to get on/off events in dhcpv4. And that's weird, because the event is added to the queue... but it's like the mgmt thread does not show up and never call the callback. I'll investigate.

tbursztyka commented:
ok I know why I don't get on/off: raising an event from an event callback is a no go. Well, if you apply the changes I requested about etherent carrier stuff, it will be "solved". But I'll still fix the net_mgmt so it should be possible to raise an event from an event callback

tbursztyka commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -1069,11 +1072,40 @@ static enum net_verdict net_dhcpv4_input(struct net_conn *conn,
+		NET_DBG("Interface %p coming up", iface);
+
+		if (iface->config.dhcpv4.state == NET_DHCPV4_RENEWING) {
+			send_request(iface);
```
I got a fix (I think) for being able to raise an event while being in an event callback.
But I get a fault at this point, I think it's too heavy from stack point of view. Do you think you could just kick in the k_delayed_work or?

jukkar commented:
Current solution needs also support from #8112 and #8113 
Lets see if we can avoid some of the net-mgmt changes.

jukkar committed:
net: ethernet: Add carrier detection to L2

Add net_eth_carrier_on() and net_eth_carrier_off() functions that
can be called by ethernet device driver when it detects that carrier
is lost or found.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Fixed version according to comments. Also added #8112 as commits in that PR are needed for this PR to work properly. This has been tested with frdm-k64f device.

tbursztyka commented on include/net/ethernet.h:
```diff
@@ -376,6 +376,51 @@ struct net_eth_hdr *net_eth_fill_header(struct ethernet_context *ctx,
+ */
+static inline void net_eth_carrier_on(struct net_if *iface)
+{
+	ethernet_mgmt_raise_carrier_on_event(iface);
```
hum that's not what I meant in my previous comment.

net_eth_carrier_on/off would be publicly declared here the same signature you did, but implemented in ethernet l2. There you would directly do what's necessary for putting if up, if carrier down etc... and raising these net mgmt event. The idea is that device drivers should not raise net mgmt event themselves (besides very specific uses case). And doing so, you could built ethernet drivers/l2 without the need of net mgmt (if reducing the memory footprint is necessary). Note that this will also reduce the code complexity in ethernet l2 (as it won't need to listen to net mgmt)

jukkar replied:
> There you would directly do what's necessary for putting if up, if carrier down etc... and raising these net mgmt event.

I was thinking that but decided to implement it like this. The reason being that I would not like to call various net_if functions from ISR context.

tbursztyka replied:
Still, as a comparision in linux it would mean relying on netlink into drivers to pass message to net core. You don't want that.

Most, if not all, ethernet drivers have their own thread. Using it would just be fine (in case mcux there is a eth_mcux_phy_event() for instance)

jukkar replied:
> Most, if not all, ethernet drivers have their own thread. Using it would just be fine (in case mcux there is a eth_mcux_phy_event() for instance)

It is impossible to say in what context the driver would call IP stack and how much stack there is for anything. In this respect the only safe way to pass this information is via mgmt event. Other option would be to use work queue for this but it looks a bit cumbersome to use.
IMHO the event passing from driver to upper stack looks quite good and convenient, I am not sure why you are so concerned about it.

tbursztyka replied:
I thought bringing the linux/netlink comparison would tell the reason. It's not "reliable". It's true that it is convenient, but logically we should not tie kind of critical system event to a user messaging event subsystem.

jukkar replied:
We can certainly use work queue here if needed but I am not buying the "reliable" thingy. The carrier ON/OFF is quite similar as network interface ON/OFF event, and if we loose the iface ON/OFF event then that is bad too.
This is a simple issue, and there is no point of arguing this. I will prepare a version that uses the work queue instead.

tbursztyka replied:
Sure I don't think there is any net core part using on/off event though. dhcpv4 is more like a user thing somehow

pfalcon commented on subsys/net/ip/net_mgmt.c:
```diff
@@ -56,7 +56,7 @@ static inline void mgmt_push_event(u32_t mgmt_event, struct net_if *iface,
 #endif /* CONFIG_NET_MGMT_EVENT_INFO */
 
-	k_sem_take(&net_mgmt_lock, K_FOREVER);
+	k_sem_take(&net_mgmt_lock, k_is_in_isr() ? K_NO_WAIT : K_FOREVER);
```
@tbursztyka, What it k_sem_take() fails with K_NO_WAIT?

tbursztyka replied:
can you comment on the original pr #8112 instead?

pfalcon replied:
Done: https://github.com/zephyrproject-rtos/zephyr/pull/8112/files#r192724235

pfalcon reviewed (commented):
Based on https://github.com/zephyrproject-rtos/zephyr/issues/7553#issuecomment-394325339 , I'd like to point that the feature of RENEWing an existing lease on interface up might be made a Kconfig-configurable feature.

(I don't argue that it *should* be such, as REQUESTing a new lease if not currently exists on interface up is definitely a good thing, and then maybe there's too little to #ifdef between the 2 cases. But thought I'd mention anyway.)

pfalcon commented:
Also IMHO this should not be stuffed into 1.12 (it should have been released on May 31th, no?).

jukkar commented:
> Also IMHO this should not be stuffed into 1.12 (it should have been released on May 31th, no?)

This is PR is fixing issue #7553 which is marked as medium bug to 1.12 thus that is why I have been working on this lately.

jukkar commented:
> This is PR is fixing issue #7553 which is marked as medium bug to 1.12 thus that is why I have been working on this lately.

I just noticed that the 1.12 target was removed from the original bug thus I will modify the target in this PR too.

nashif commented:
> I just noticed that the 1.12 target was removed from the original bug thus I will modify the target in this PR too.

bugs usually do not have a target by default, bugs apply to the coming release unless we set something to the following release, so this is still valid for 1.12.

jukkar commented:
Updated the PR so that ethernet driver uses work queue instead of generating the carrier on/off event itself.

jukkar commented:
Rebased against master.

jukkar commented:
Changed the dhcpv4 handling and removed unnecessary checks.

pfalcon commented:
Hope also @therealprof can confirm and ack that this PR in its current state works for him. (I don't have an advanced enough DHCP setup to test it myself for example.)

pfalcon commented:
@therealprof 

> Can't test this since the STM32 driver does not generate the required events. 

D'oh, sorry!

> I've seen one implementation where they essentially periodically polled the PHY for the current link status but meh...

That's how frdm_k64f, as used in this PR, is implemented, so don't be too shy about it - nice features sometimes require some workarounds/overhead ;-).

tbursztyka reviewed (changes requested):
some ifdefs are not required anymore, rest is fine to me.

jukkar commented on subsys/net/ip/l2/ethernet/ethernet.c:
```diff
@@ -757,18 +759,74 @@ int net_eth_vlan_disable(struct net_if *iface, u16_t tag)
 NET_L2_INIT(ETHERNET_L2, ethernet_recv, ethernet_send, ethernet_reserve,
 	    ethernet_enable);
 
+#if defined(CONFIG_NET_L2_ETHERNET_MGMT)
```
> You don't need that actually. Now that you added a patch "stubbing" ethernet_mgmt_raise_carrier_on/off_event() it will build with or w/o net_mgmt.

Actually I did not had these checks here because of that. My reasoning was that net_eth_carrier_on() would be no-op if ethernet mgmt is turned off, but I suppose we can have that function to work regardless of mgmt support.

tbursztyka replied:
yep, net_eth_carrier_on/off are not mgmt dependent. If mgmt is enabled it will raise the mgmt even (for whoever is listening to these), but that's all.

jukkar commented:
Updated according to comments.

tbursztyka commented on include/net/ethernet.h:
```diff
@@ -21,6 +21,7 @@
 #include <net/net_pkt.h>
 #include <misc/util.h>
 #include <net/net_if.h>
+#include <net/net_mgmt.h>
```
You don't need it anymore I guess

tbursztyka reviewed (approved):
tiny comment, rest is fine
